### PR TITLE
Advanced Building Blocks - Enumerables

### PR DESCRIPTION
**Pull request by DcRonan**
###  Enumerable Methods

1. Created our own solutions to the Ruby methods.
2. Used 'yield'
3. Used procs and blocks
3. used Ruby, Rubocop, irb and repl.it
4. test file for test cases.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+| Contents        |
+| ------------- |
+| [Built With](#built-with) |
+| [Live Demo](#live-demo) |
+| [Run Tests](#run-tests) |
+| [Authors](#authors) | 
+| [License](#license) | 
+
+## Description
+> In this project we was given the task to create identical methods that 'Ruby' provides in our own solutions.
```
Comment (kitpao) [inline README.md:14]: - [x] **OPTIONAL** Just a typo here: instead of We was given -> We were given...

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @DcRonan and @chelmerrox 

🌟 I really liked your Readme. Also, good job following Git Flow 🍏 

**_Note:_** All the **[OPTIONAL]** are **not required** in order to approve your project but are personal suggestions to improve your coding experience

- [x] Your commits are descriptive but try following the same structure for them: remember to write your commits in the imperative tense, as stated in [the article pointed in pathwright](https://www.freecodecamp.org/news/level-you-up-to-awesome-commit-messages-a85558cb90e8/)

### Your methods
- [x] For all methods, the reason your methods do not work inside the enumerable module is because of the format. You don't need to specify an argument since enumerables already take them like this: argument.method (as in `[1,2,3].my_each`). The only method that has to be outside the enumerable module is #multiply_els(argument). So for the rest of the methods definitions you can use it as:
  ```
  def my_each
    # some code, use "self" to refer to the argument
  end
  ```
 - [x] For all methods that do not work without a block, please return the enumerable item `return to_enum(:name_of_your_method)`, instead of returning the argument
- [x] The methods that work with arrays, should also work with ranges

👉 Please correct these issues first so we can help you with the rest, it seems you have some trouble in the logic of the methods from my_all onwards, but I cannot say from certain right now.

You can check how the original methods work in this link: https://ruby-doc.org/core-2.6.5/Enumerable.html

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me in slack as @kitpao

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      my_each { |x| return result unless x == arg }
+    elsif !arg.nil? && arg.is_a?(Regexp) || arg.is_a?(String)
+      my_each { |y| return result unless y.match(arg) }
+    elsif !block_given?
+      my_each { |a| return result if a.nil? }
+    elsif block_given?
+      my_each { |b| return result unless yield(b) }
+    end
+    !result
+  end
```
Comment (BrittanyBlake) [inline main.rb]: - [ ] If no block or argument is given, `my_all` should return false when one of the members in the collection are false or nil.
```ruby 
 [ "hey", false, true ].all? == [ false, true ].my_all? => false
```

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+      my_each { |x| return result if x == arg }
+    elsif !arg.nil? && arg.is_a?(Regexp) || arg.is_a?(String)
+      my_each { |y| return result if y.match(arg) }
+    elsif !block_given?
+      my_each { |a| return result unless a.nil? }
+    elsif block_given?
+      my_each { |b| return result if yield(b) }
+    end
+    !result
+  end
```
Comment (BrittanyBlake) [inline main.rb]: - [ ] If no block or argument is given, `my_any` should return false if at least one of the collection is not true

```ruby 
 [ nil, false, false ].any? == [ nil, false, false ].my_any? => false
```

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    end
+
+    if block_given? && !Array(self).nil?
+      arr_size.times do |i|
+        count += 1 if yield(Array(self)[i])
+      end
+    end
+    count
+  end
+
```
Comment (BrittanyBlake) [inline main.rb]: - [x] if an argument is given, `my_count` should count the number of items in an enumerator that are equal to item. 
```ruby
 array = [0,0,3,0,5,6,8,7,9]
  array.count(0) ===  array.my_count(0) => false
```

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+  def my_map(&_proc)
+    return to_enum unless block_given?
+
+    new_arr = []
+    my_each do |i|
+      new_arr.push(yield(i))
+    end
+    new_arr
+  end
+
```
Comment (BrittanyBlake) [inline main.rb]: - [ ] When both a proc and a block are given, `my_map` should only execute the proc and it should return a new array with the results of running the proc once for every element in enum.
* Yours is currently returning an argument error. Remember that the `map` method can be used without an argument. You can read more about it [here](https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-map)

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    return unless block_given?
+
+    return unless !accumulator.nil? && !operator.is_a?(Symbol) || !operator.is_a?(Integer)
+
+    arr_size.times do |i|
+      return result += yield(sum, Array(self)[i]) if Array(self)[i].class == Integer
+    end
+  end
+end
```
Comment (BrittanyBlake) [inline main.rb]: - [x] `my_inject` should **raise** a `localJumpError` when no block or argument is given. Yours is not currently raising any errors.

- [x] It should also search for the longest word in an array of strings and return that **word**. Yours currently returns an integer.

- [x] When a block is given without an initial value, `my_inject` should combine all of the elements of enum by applying a binary operation that is specified by a block.
```ruby
[5, 2, 1].inject { |result, x| result + x } == [5, 2, 1].my_inject { |result, x| result + x }  => false
```
- [x] When `my_inject` is called on a range and a block is given without an initial value, it should combine also all of the elements of enum by applying a binary operation that is specified by a block.

- [x] When a block is given with an initial value, `my_inject` should combine all of the elements of enum by applying a binary operation that is specified by a block.
```ruby
 [5, 2, 1].inject(0) { |result, x| result + x } ==  [5, 2, 1].my_inject(0) { |result, x| result + x }  => false
```

- [x] When a symbol is specified without an initial value, `my_inject` should combine each of the elements of the collection by applying the symbol as a named method.

- [x] When `my_inject` is called on a range, and a symbol is specified without an initial value, it should combine each of the elements of the collection by applying the symbol as a named method.

- [x] When a symbol is specified with an initial value, `my_inject` should combine each of the elements of the collection by applying the symbol as a named method.

- [x] When `my_inject` is called on a range, and a symbol is specified with an initial value, it should combine each of the elements of the collection by applying the symbol as a named method.

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+# =============
+#  multiply_els
+# =============
+
+def multiply_els(arr)
+  arr.my_inject(:*)
+end
```
Comment (BrittanyBlake) [inline main.rb:189]: - [x] Multiply els should accept an array as an argument and multiply all of the elements of the array together using `#my_inject`. 

- This should work once the errors in `#my_inject` are solved.

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]
### Status: Changes Required ♻️ 
****
### Hi @chelmerrox and @DcRonan ! 👋 
You have done a great job implementing the changes requested by the previous TSE! 👍 You have also done a wonderful job with your methods so far! 👏 👏 👏  There are just a few issues that need to be resolved before you can move on: ⬇️ 
****
_FYI: Please note that it is not necessary to commit the file with your test cases_
****
### Please make the required changes and resubmit for a new code review once you are ready 🙏


Good Luck and Happy Coding 🍀 👩🏾‍💻

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Brittany Blake)._

File: main.rb (hunk 2 above)
Comment (DcRonan) [reply main.rb]: This already returns false.

File: main.rb (hunk 5 above)
Comment (DcRonan) [reply main.rb]: This isn't possible to return just a block when a proc and a block is given as a proc is a reusable block.

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,
Thank you for applying  the suggested changes
There are still one bug to fix
```rb
2.6.5 :014 > array
 => [1, 2, 5, 2, 3, 2, 1, 3, 8, 6, 4, 4, 2, 8, 1, 1, 1, 0, 2, 4, 3, 8, 5, 7, 5, 5, 6, 6, 3, 0, 3, 6, 4, 7, 6, 4, 0, 4, 1, 0, 0, 0, 7, 7, 3, 8, 5, 5, 4, 5, 4, 1, 0, 3, 3, 0, 7, 8, 1, 2, 2, 1, 1, 8, 6, 6, 1, 7, 6, 5, 1, 0, 4, 1, 1, 2, 0, 6, 2, 1, 2, 2, 3, 2, 2, 5, 1, 4, 2, 5, 3, 4, 3, 7, 4, 0, 3, 7, 8, 5] 
2.6.5 :015 > my_proc = proc {|num| num > 10 }
2.6.5 :016 > array.my_map(my_proc) {|num| num < 10 }
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):16
        1: from main.rb:142:in `my_map'
ArgumentError (wrong number of arguments (given 1, expected 0))
2.6.5 :017 > array.map(&my_proc)
 => [false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false] 
2.6.5 :018 > 
```

`array.my_map(my_proc) {|num| num < 10 }` must executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

Kindly fix it and submit a new code review request
Thanks

Mael FOSSO

**Comment by DcRonan** [comment]
Haven't made any changes as the last review was incorrect.

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Kindly read  that 
> 14. Modify your #my_map method to take either a proc or a block. It won't be necessary to apply both a proc and a block in the same #my_map call since you could get the same effect by chaining together one #my_map call with the block and one with the proc. This approach is also clearer, since the user doesn't have to remember whether the proc or block will be run first. So if both a proc and a block are given, only execute the proc.

From https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods

Once done, implement it correctly, please.

Mael FOSSO

**Comment by DcRonan** [comment]
This isn't possible so please can you stop asking me to do it. I have checked your code and it returns the exact same as what my code returns. For the next TSE to review this, please can you check your own to see if your code returns what it asks. Thanks.

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]
Hello @DcRonan,
 Your `my_map` method is failing. 
Screen Shot 2020-09-03 at 11 41 54 PM


- Please modify your method so that if a block and a proc and both given, it will only execute the proc. This is a requirement and it would not be asked of you if it were not possible. 

Please see the requirements below 👇 

> Modify your #my_map method to take either a proc or a block. It won't be necessary to apply both a proc and a block in the same #my_map call since you could get the same effect by chaining together one #my_map call with the block and one with the proc. This approach is also clearer, since the user doesn't have to remember whether the proc or block will be run first. So if both a proc and a block are given, only execute the proc.

or you can read it [here](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methodss). 

- [x] Additionally, please ensure that your `my_inject` method is raising a `localJumpError` if no block or argument is given.

Once you have implemented the required changes, you can resubmit for a review.

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        arg_one = yield(arg_one, i)
+      end
+    else
+      arr.my_each do |i|
+        arg_one = arg_one.send(arg_two, i)
+      end
+    end
+    arg_one
+  end
+end
```
Comment (lekegitrepo) [inline main.rb]: - **_`#my_inject` fails the following test cases_**
**_`array = [1, 2, 5, 2, 3, 2, 1, 3, 8, 6, 4, 4, 2, 8, 1, 1, 1, 0, 2, 4, 3, 8, 5, 7, 5, 5, 6, 6, 3, 0, 3, 6, 4, 7, 6, 4, 0, 4, 1, 0, 0, 0, 7, 7, 3, 8, 5, 5, 4, 5, 4, 1, 0, 3, 3, 0, 7, 8, 1, 2, 2, 1, 1, 8, 6, 6, 1, 7, 6, 5, 1, 0, 4, 1, 1, 2, 0, 6, 2, 1, 2, 2, 3, 2, 2, 5, 1, 4, 2, 5, 3, 4, 3, 7, 4, 0, 3, 7, 8, 5]`_**

**_`range = Range.new(5, 50)`_**

#### Using rspec for the test
**_1. `expect { array.my_inject }.to raise_error(LocalJumpError)`_**

**_2. `expect(array.my_inject(:+)).to eq array.inject(:+)`_**

**_3. `expect(range.my_inject(:*)).to eq range.inject(:*)`_**

**_4. `expect(array.my_inject(20, :*)).to eq array.inject(20, :*)`_**

**_5. `expect(range.my_inject(2, :*)).to eq range.inject(2, :*)`_**

**_6. `actual = multiply_els [2, 4, 5]      expect(actual).to eq 40`_**

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
### Status: Request changes :recycle:
#### Hi @DcRonan and @chelmerrox :clap: :clap:
#### You have done a good job, however, some test cases are still failing. I know that you have put a lot of effort to implement the required changes, that is why I have added the snippet for the test cases that are failing so that you can check and confirm it yourself. Check my comments below :point_down:

#### _Reviewed by Temitayo Adeleke._

File: main.rb (hunk 8 above)
Comment (DcRonan) [reply main.rb]: fixed and tested all your test cases, they now all pass.

**Comment by Dandush03** [comment]
# Status: :red_circle: Require Changes :red_circle: 
### You're almost done! :partying_face: 
There still 2 cases that are failing. and I will point out below.
### #multiply_els
- [x] raises a "LocalJumpError" when no block or argument is given
- [x] accepts an array as an argument and multiplies all the elements of the array together using #my_inject
### I know it's some time dificult and frustrating but all details are important when you code to avoid an error. you just need to add an IF case and you will be good to go
# Happy Coding :computer: :nerd_face:

**Comment by DcRonan** [comment]
@Dandush03 Hey, my_inject already raises the "LocalJumpError" when no block or argument is given and I have fixed the other thing you requested.

**Review by addod19 — APPROVED** [review]
## Status: Approved :trophy: :balloon: 

Hi **TEAM**, Congratulations for the hard work you put into this project :+1: Thanks to your persistent hard work and the help of our TSE's, now our project meets all the minimum requirements :fireworks: Hurray!!!

Go ahead and merge into development :smile: 
Do not forget to submit the project completion form :100: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo/)

**Pull request merged**